File: examples/19_checkpoint_auto_resume.py
"""Example 19: Checkpoint Auto-Resume — preserve upstream work on schema changes.

Scenario: A 4-node linear pipeline (prepare -> enrich -> analyze -> report)
processes data. After the first run completes, the analyze node's output
model changes (a field is added). On the second run with the same thread_id,
neograph detects the schema divergence and automatically rewinds to the
checkpoint before the changed node. Nodes A and B are preserved from the
checkpoint; nodes C and D re-execute.

This demonstrates two behaviors:
  1. auto_resume=True (default) — automatic rewind and selective re-execution
  2. auto_resume=False — raises CheckpointSchemaError so you can handle it

Patterns demonstrated:
  - Scripted @node pipeline (no LLM, no API keys)
  - MemorySaver checkpointer for in-memory checkpoint persistence
  - construct_from_functions for pipeline assembly
  - Schema fingerprinting and per-node invalidation

Run:
    python examples/19_checkpoint_auto_resume.py
"""

from __future__ import annotations

from pydantic import BaseModel

from langgraph.checkpoint.memory import MemorySaver

from neograph import (
    CheckpointSchemaError,
    compile,
    construct_from_functions,
    node,
    run,
)


# -- Schemas ------------------------------------------------------------------
# Shared types that stay the same across both pipeline versions.

class Prepared(BaseModel):
    text: str
    token_count: int


class Enriched(BaseModel):
    text: str
    token_count: int
    entities: list[str]


class Report(BaseModel):
    title: str
    body: str


# -- Demo 1: auto_resume=True (default) — selective re-execution -------------

def demo_auto_resume():
    """Automatic rewind and selective re-execution on schema change."""
    print("=" * 60)
    print("DEMO 1: auto_resume=True (default)")
    print("=" * 60)

    # Analysis output types — v1 and v2 differ by one field.
    # Defined locally so the checkpoint serializer round-trips them as dicts,
    # allowing the new state model to absorb old checkpoint data.

    class AnalysisV1(BaseModel):
        summary: str
        entity_count: int

    class AnalysisV2(BaseModel):
        summary: str
        entity_count: int
        confidence: float = 0.0  # new field in v2

    # -- v1 nodes --

    @node(outputs=Prepared)
    def prepare() -> Prepared:
        print("  [prepare] running")
        text = "The system validates all user inputs and logs access attempts."
        return Prepared(text=text, token_count=len(text.split()))

    @node(outputs=Enriched)
    def enrich(prepare: Prepared) -> Enriched:
        print("  [enrich] running")
        entities = ["system", "user", "inputs", "access"]
        return Enriched(
            text=prepare.text,
            token_count=prepare.token_count,
            entities=entities,
        )

    @node(outputs=AnalysisV1, name="analyze")
    def analyze_v1(enrich: Enriched) -> AnalysisV1:
        print("  [analyze] running (v1)")
        return AnalysisV1(
            summary=f"Found {len(enrich.entities)} entities in {enrich.token_count} tokens",
            entity_count=len(enrich.entities),
        )

    @node(outputs=Report, name="report")
    def report_v1(analyze: AnalysisV1) -> Report:
        print("  [report] running (v1)")
        return Report(title="Analysis Report", body=analyze.summary)

    # -- v2 nodes (analyze output gains a field) --

    @node(outputs=AnalysisV2, name="analyze")
    def analyze_v2(enrich: Enriched) -> AnalysisV2:
        print("  [analyze] running (v2 -- new field)")
        return AnalysisV2(
            summary=f"Found {len(enrich.entities)} entities in {enrich.token_count} tokens",
            entity_count=len(enrich.entities),
            confidence=0.92,
        )

    @node(outputs=Report, name="report")
    def report_v2(analyze: AnalysisV2) -> Report:
        print("  [report] running (v2)")
        return Report(
            title="Analysis Report v2",
            body=f"{analyze.summary} (confidence={analyze.confidence})",
        )

    checkpointer = MemorySaver()
    config = {"configurable": {"thread_id": "demo-auto-resume"}}

    # -- Run 1: full pipeline, all 4 nodes execute --
    print("\nRun 1: full pipeline (v1)")
    print("-" * 40)
    pipeline_v1 = construct_from_functions(
        "checkpoint-demo", [prepare, enrich, analyze_v1, report_v1],
    )
    graph_v1 = compile(pipeline_v1, checkpointer=checkpointer)
    result_v1 = run(graph_v1, input={"node_id": "demo"}, config=config)
    print(f"\n  Result: {result_v1['report'].body}")

    # -- Run 2: schema changed on analyze, only analyze+report re-execute --
    print("\nRun 2: analyze output changed (v1 -> v2)")
    print("-" * 40)
    print("  Expected: prepare and enrich preserved, analyze and report re-run")
    pipeline_v2 = construct_from_functions(
        "checkpoint-demo", [prepare, enrich, analyze_v2, report_v2],
    )
    graph_v2 = compile(pipeline_v2, checkpointer=checkpointer)
    result_v2 = run(graph_v2, input={"node_id": "demo"}, config=config)
    print(f"\n  Result: {result_v2['report'].body}")


# -- Demo 2: auto_resume=False — CheckpointSchemaError -----------------------

def demo_strict_mode():
    """Strict mode: raise CheckpointSchemaError on mismatch."""
    print("\n" + "=" * 60)
    print("DEMO 2: auto_resume=False (strict mode)")
    print("=" * 60)

    class AnalysisV1(BaseModel):
        summary: str
        entity_count: int

    class AnalysisV2(BaseModel):
        summary: str
        entity_count: int
        confidence: float = 0.0

    @node(outputs=Prepared, name="prepare")
    def prepare_s() -> Prepared:
        print("  [prepare] running")
        text = "The system validates all user inputs and logs access attempts."
        return Prepared(text=text, token_count=len(text.split()))

    @node(outputs=Enriched, name="enrich")
    def enrich_s(prepare: Prepared) -> Enriched:
        print("  [enrich] running")
        entities = ["system", "user", "inputs", "access"]
        return Enriched(
            text=prepare.text,
            token_count=prepare.token_count,
            entities=entities,
        )

    @node(outputs=AnalysisV1, name="analyze")
    def analyze_s_v1(enrich: Enriched) -> AnalysisV1:
        print("  [analyze] running (v1)")
        return AnalysisV1(
            summary=f"Found {len(enrich.entities)} entities",
            entity_count=len(enrich.entities),
        )

    @node(outputs=Report, name="report")
    def report_s_v1(analyze: AnalysisV1) -> Report:
        print("  [report] running")
        return Report(title="Report", body=analyze.summary)

    @node(outputs=AnalysisV2, name="analyze")
    def analyze_s_v2(enrich: Enriched) -> AnalysisV2:
        print("  [analyze] running (v2)")
        return AnalysisV2(
            summary=f"Found {len(enrich.entities)} entities",
            entity_count=len(enrich.entities),
            confidence=0.95,
        )

    @node(outputs=Report, name="report")
    def report_s_v2(analyze: AnalysisV2) -> Report:
        print("  [report] running (v2)")
        return Report(title="Report v2", body=analyze.summary)

    checkpointer = MemorySaver()
    config = {"configurable": {"thread_id": "demo-strict"}}

    # -- Run 1: full pipeline --
    print("\nRun 1: full pipeline (v1)")
    print("-" * 40)
    pipeline_v1 = construct_from_functions(
        "strict-demo", [prepare_s, enrich_s, analyze_s_v1, report_s_v1],
    )
    graph_v1 = compile(pipeline_v1, checkpointer=checkpointer)
    run(graph_v1, input={"node_id": "demo"}, config=config)

    # -- Run 2: schema changed, strict mode rejects --
    print("\nRun 2: schema changed, auto_resume=False")
    print("-" * 40)
    pipeline_v2 = construct_from_functions(
        "strict-demo", [prepare_s, enrich_s, analyze_s_v2, report_s_v2],
    )
    graph_v2 = compile(pipeline_v2, checkpointer=checkpointer)
    try:
        run(graph_v2, input={"node_id": "demo"}, config=config, auto_resume=False)
        print("  ERROR: expected CheckpointSchemaError but none was raised")
    except CheckpointSchemaError as e:
        print(f"  Caught CheckpointSchemaError (expected)")
        print(f"  Invalidated nodes: {sorted(e.invalidated_nodes)}")
        print(f"  Message: {e}")


if __name__ == "__main__":
    # Suppress framework logs so the example output is clear.
    import logging
    import structlog
    logging.getLogger().setLevel(logging.WARNING)
    structlog.configure(wrapper_class=structlog.make_filtering_bound_logger(logging.WARNING))

    demo_auto_resume()
    demo_strict_mode()

File: docs/design/ts-parity/19_checkpoint_auto_resume.ts
// TS parity sketch of: examples/19_checkpoint_auto_resume.py
// NOT runnable. Hypothetical port against the PROPOSED API in
// docs/design/typescript-port.md (AD-0 transformer + .pipe + node() wrapper).
//
// What this example exercises (all keyless/scripted — no LLM, no DI, no Oracle/Each/Loop,
// no ForwardConstruct, no describe_type):
//   - scripted @node pipeline (4 linear nodes: prepare -> enrich -> analyze -> report)
//   - construct_from_functions assembly
//   - compile(..., checkpointer=MemorySaver())
//   - run(..., config={thread_id}, auto_resume=True|False)
//   - schema fingerprinting + per-node invalidation + auto-rewind on schema change
//   - CheckpointSchemaError.invalidated_nodes
//   - locally-scoped output models whose checkpoint serde round-trips as dicts
//
// The DX-hard part of THIS example is NOT node authoring — it is the checkpoint
// schema-fingerprint auto-rewind subsystem, which the parity matrix does not cover.

import { z } from "zod";
import { MemorySaver } from "@langchain/langgraph";
import {
  node,
  compile,
  run,
  constructFromFunctions,
  CheckpointSchemaError,
} from "@neograph/core";

// -- Schemas -----------------------------------------------------------------
// PARITY(direct): Pydantic BaseModel -> Zod object. Attribute access
// `result.report.body` in Python (a model INSTANCE) becomes property access on a
// Zod-parsed plain object — same call site. AD-1 (Zod) row.

const Prepared = z.object({ text: z.string(), tokenCount: z.number().int() });
type Prepared = z.infer<typeof Prepared>;

const Enriched = z.object({
  text: z.string(),
  tokenCount: z.number().int(),
  entities: z.array(z.string()),
});
type Enriched = z.infer<typeof Enriched>;

const Report = z.object({ title: z.string(), body: z.string() });
type Report = z.infer<typeof Report>;

// -- Demo 1: autoResume=true (default) — selective re-execution --------------

async function demoAutoResume() {
  console.log("=".repeat(60));
  console.log("DEMO 1: autoResume=true (default)");
  console.log("=".repeat(60));

  // Analysis output types — v1 and v2 differ by one field.
  //
  // PARITY(redesign): in Python these are defined *inside the function* so the
  // checkpoint serializer round-trips them as plain dicts and the new state model
  // can absorb old checkpoint data. In TS there are no runtime classes to begin
  // with — state holds plain objects and Zod .parse() rehydrates them. The
  // "absorb old checkpoint" behavior maps to Zod: the new v2 field carries
  // `.default(0)`, so an old checkpoint dict lacking `confidence` still parses.
  // (Python's `confidence: float = 0.0` default -> `z.number().default(0)`.)
  const AnalysisV1 = z.object({ summary: z.string(), entityCount: z.number().int() });
  const AnalysisV2 = z.object({
    summary: z.string(),
    entityCount: z.number().int(),
    confidence: z.number().default(0), // new field in v2
  });
  type AnalysisV1 = z.infer<typeof AnalysisV1>;
  type AnalysisV2 = z.infer<typeof AnalysisV2>;

  // -- v1 nodes --
  // PARITY(direct, via AD-0): the transformer reads the fn signature and emits
  // __neo_meta.inputs from parameter names + __neo_meta.output from the return
  // type — so `enrich(prepare: Prepared)` wires the prepare->enrich edge exactly
  // like Python's inspect.signature. `outputs:`/`name:` are explicit config.
  //
  // PARITY(gap): the AD-0 wrapper `node({...}, (prepare: Prepared): Enriched => ...)`
  // cannot recover the *return type* the way Python reads `-> Enriched`, because a
  // scripted body may `return` at several points. The transformer must trust the
  // arrow's declared return annotation. Where it can't, fall back to explicit
  // `outputs:` (shown here for robustness, since checkpoint fingerprinting below
  // depends on this schema being exact).

  const prepare = node({ outputs: Prepared }, (): Prepared => {
    console.log("  [prepare] running");
    const text = "The system validates all user inputs and logs access attempts.";
    return { text, tokenCount: text.split(" ").length };
  });

  const enrich = node({ outputs: Enriched }, (prepare: Prepared): Enriched => {
    console.log("  [enrich] running");
    return {
      text: prepare.text,
      tokenCount: prepare.tokenCount,
      entities: ["system", "user", "inputs", "access"],
    };
  });

  // PARITY(direct): `name: "analyze"` overrides the wrapper-inferred name so both
  // schema versions occupy the SAME node identity across runs. This is what makes
  // the fingerprint of state field `analyze` change (not add a new field).
  const analyzeV1 = node({ outputs: AnalysisV1, name: "analyze" }, (enrich: Enriched): AnalysisV1 => {
    console.log("  [analyze] running (v1)");
    return {
      summary: `Found ${enrich.entities.length} entities in ${enrich.tokenCount} tokens`,
      entityCount: enrich.entities.length,
    };
  });

  const reportV1 = node({ outputs: Report, name: "report" }, (analyze: AnalysisV1): Report => {
    console.log("  [report] running (v1)");
    return { title: "Analysis Report", body: analyze.summary };
  });

  // -- v2 nodes (analyze output gains a field) --
  const analyzeV2 = node({ outputs: AnalysisV2, name: "analyze" }, (enrich: Enriched): AnalysisV2 => {
    console.log("  [analyze] running (v2 -- new field)");
    return {
      summary: `Found ${enrich.entities.length} entities in ${enrich.tokenCount} tokens`,
      entityCount: enrich.entities.length,
      confidence: 0.92,
    };
  });

  const reportV2 = node({ outputs: Report, name: "report" }, (analyze: AnalysisV2): Report => {
    console.log("  [report] running (v2)");
    return {
      title: "Analysis Report v2",
      body: `${analyze.summary} (confidence=${analyze.confidence})`,
    };
  });

  const checkpointer = new MemorySaver();
  const config = { configurable: { thread_id: "demo-auto-resume" } };

  // -- Run 1: full pipeline, all 4 nodes execute --
  console.log("\nRun 1: full pipeline (v1)");
  // PARITY(direct): construct_from_functions -> constructFromFunctions. The
  // transformer-emitted __neo_meta on each node lets the assembler build the DAG
  // from param names, same as the Python builder walking sidecars.
  const pipelineV1 = constructFromFunctions("checkpoint-demo", [prepare, enrich, analyzeV1, reportV1]);
  const graphV1 = compile(pipelineV1, { checkpointer });
  const resultV1 = await run(graphV1, { input: { node_id: "demo" }, config });
  console.log(`\n  Result: ${(resultV1.report as Report).body}`);

  // -- Run 2: schema changed on analyze, only analyze+report re-execute --
  //
  // PARITY(GAP — the whole point of this example is unmapped): neograph rewinds
  // here because compile() stashed graph.schema_fingerprint (compute_schema_fingerprint,
  // state.py:401) + graph.node_fingerprints (compute_node_fingerprints, state.py:346),
  // and on resume runner.py diffs them, computes `invalidated = {analyze, report}`,
  // walks get_state_history() backwards for the oldest checkpoint whose `.next`
  // intersects `invalidated`, injects that checkpoint_id, and invoke(None). The
  // parity matrix has NO row for any of this — schema fingerprint, node fingerprint,
  // _type_signature, auto-rewind, get_state_history divergence walk. See report.
  console.log("\nRun 2: analyze output changed (v1 -> v2)");
  console.log("  Expected: prepare and enrich preserved, analyze and report re-run");
  const pipelineV2 = constructFromFunctions("checkpoint-demo", [prepare, enrich, analyzeV2, reportV2]);
  const graphV2 = compile(pipelineV2, { checkpointer });
  const resultV2 = await run(graphV2, { input: { node_id: "demo" }, config });
  console.log(`\n  Result: ${(resultV2.report as Report).body}`);
}

// -- Demo 2: autoResume=false — CheckpointSchemaError ------------------------

async function demoStrictMode() {
  console.log("\n" + "=".repeat(60));
  console.log("DEMO 2: autoResume=false (strict mode)");
  console.log("=".repeat(60));

  const AnalysisV1 = z.object({ summary: z.string(), entityCount: z.number().int() });
  const AnalysisV2 = z.object({
    summary: z.string(),
    entityCount: z.number().int(),
    confidence: z.number().default(0),
  });
  type AnalysisV1 = z.infer<typeof AnalysisV1>;
  type AnalysisV2 = z.infer<typeof AnalysisV2>;

  const prepare = node({ outputs: Prepared, name: "prepare" }, (): Prepared => {
    const text = "The system validates all user inputs and logs access attempts.";
    return { text, tokenCount: text.split(" ").length };
  });
  const enrich = node({ outputs: Enriched, name: "enrich" }, (prepare: Prepared): Enriched => ({
    text: prepare.text,
    tokenCount: prepare.tokenCount,
    entities: ["system", "user", "inputs", "access"],
  }));
  const analyzeV1 = node({ outputs: AnalysisV1, name: "analyze" }, (enrich: Enriched): AnalysisV1 => ({
    summary: `Found ${enrich.entities.length} entities`,
    entityCount: enrich.entities.length,
  }));
  const reportV1 = node({ outputs: Report, name: "report" }, (analyze: AnalysisV1): Report => ({
    title: "Report",
    body: analyze.summary,
  }));
  const analyzeV2 = node({ outputs: AnalysisV2, name: "analyze" }, (enrich: Enriched): AnalysisV2 => ({
    summary: `Found ${enrich.entities.length} entities`,
    entityCount: enrich.entities.length,
    confidence: 0.95,
  }));
  const reportV2 = node({ outputs: Report, name: "report" }, (analyze: AnalysisV2): Report => ({
    title: "Report v2",
    body: analyze.summary,
  }));

  const checkpointer = new MemorySaver();
  const config = { configurable: { thread_id: "demo-strict" } };

  console.log("\nRun 1: full pipeline (v1)");
  const pipelineV1 = constructFromFunctions("strict-demo", [prepare, enrich, analyzeV1, reportV1]);
  const graphV1 = compile(pipelineV1, { checkpointer });
  await run(graphV1, { input: { node_id: "demo" }, config });

  console.log("\nRun 2: schema changed, autoResume=false");
  const pipelineV2 = constructFromFunctions("strict-demo", [prepare, enrich, analyzeV2, reportV2]);
  const graphV2 = compile(pipelineV2, { checkpointer });
  try {
    // PARITY(direct, given the subsystem exists): the auto_resume flag is a plain
    // run() option. The redesign is entirely under the hood in what raises this.
    await run(graphV2, { input: { node_id: "demo" }, config, autoResume: false });
    console.log("  ERROR: expected CheckpointSchemaError but none was raised");
  } catch (e) {
    if (e instanceof CheckpointSchemaError) {
      // PARITY(direct): error class + `invalidatedNodes` field. Python exposes a
      // set; TS uses a Set<string>. Error hierarchy row = Direct.
      console.log("  Caught CheckpointSchemaError (expected)");
      console.log(`  Invalidated nodes: ${[...e.invalidatedNodes].sort()}`);
      console.log(`  Message: ${e.message}`);
    } else {
      throw e;
    }
  }
}

async function main() {
  await demoAutoResume();
  await demoStrictMode();
}

void main();
